Update EIP-7819: Add trailing data to the delegation designator (#12047)
ethereum/EIPs: EIPS/eip-7819.md

and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).

A new instruction (`SETDELEGATE`) is added at `0xf6`. It takes four stack operands and produces one:

```
SETDELEGATE
  stack input (top first):  salt, target, data_offset, data_len
  stack output:             location
```

### Behavior

Executing this instruction does the following:

1. Deduct `EMPTY_ACCOUNT_COST` gas.
2. Halt if the current frame is in `static-mode`.
3. Pop `salt`, `target`, `data_offset`, and `data_len` from the operand stack.
4. Halt if `data_len` is greater than `MAX_DATA_LEN`
5. Charge gas for memory expansion gas for the region `[data_offset, data_offset + data_len)` as for any memory read
6. Charge gas for code data costs of `DATA_WORD_COST * ceil(data_len / 32)`.
7. Calculate `location` as `keccak256(DESIGNATOR ++ address ++ salt)[12:]`.
8. Add `location` to `accessed_addresses`, as defined in [EIP-2929](./eip-2929.md).
9. Halt if the code at `location` is not empty and does not start with b`DESIGNATOR`.
10. Add `EMPTY_ACCOUNT_COST - BASE_COST` gas to the global refund counter if `location` exists in the trie.

> 11. Set the code of `location` to either `DESIGNATOR ++ target ++ data or empty, matching the delegation process defined in EIP-7702,

Reviewer: ```suggestion
11. Set the code of `location` to either `DESIGNATOR ++ target ++ data` or empty, matching the delegation process defined in EIP-7702,
```

    - `data` is `memory[data_offset : data_offset + data_len]`
    - If `data_len` is 0, the code written is exactly `DESIGNATOR ++ target`.
    - Similarly to EIP-7702, if `target` is `0x0` (regardless of `data_len`), do not write the designation. Instead, clear the code at `location` and reset the `location`'s code hash to the empty hash `0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470`.
12. If the nonce of the account at `location` is 0, increment it to 1
13. push `location` onto the stack

> In step 9, `target` MUST be exactly 20 bytes. If the value retrieved from the stack is smaller than 20 bytes, it MUST be left-padded (big-endian notation) with zeros. If it is larger than 20 bytes, the highest-order bytes MUST be trimmed. The code placed at `location` MUST be exactly 23 bytes long, following EIP-7702.

Reviewer: ```suggestion
In step 11, `target` MUST be exactly 20 bytes. If the value retrieved from the stack is smaller than 20 bytes, it MUST be left-padded (big-endian notation) with zeros. If it is larger than 20 bytes, the highest-order bytes MUST be trimmed. The code placed at `location` MUST be exactly 23 + data_len bytes long.
```

> The write is not conditioned on the previous contents of `location` beyond the step 8 check: a designator MAY be replaced by a designator with a different target, different trailing data, or both. Code that is not a designator is never overwritten.

Reviewer: ```suggestion
The write is not conditioned on the previous contents of `location` beyond the step 9 check: a designator MAY be replaced by a designator with a different target, different trailing data, or both. Code that is not a designator is never overwritten.
```

Step 12 ensures that an account created using `SETDELEGATE` never returns to an empty state (as forbidden by [EIP-7523](./eip-7523.md)). Even if the balance is zero, no state was ever set, and the code is removed by a `SETDELEGATE` operation with a zero target, the nonce being at least one guarantees the account is not considered empty.

Delegated code reads its account's trailing data from `code[23:]` using `EXTCODECOPY(ADDRESS, …)` with length `EXTCODESIZE(ADDRESS) - 23`. It MUST use the `EXTCODE*` family against `ADDRESS` rather than `CODESIZE` and `CODECOPY`: during delegated execution the plain `CODE*` instructions observe the executing target's own code, so they see neither the designator nor the trailing data.

### Parameters

The first three parameters are the same ones used in EIP-7702.

| Constant             | Value    |
| -------------------- | -------- |
| `DESIGNATOR`         | 0xef0100 |
| `EMPTY_ACCOUNT_COST` | 25000    |
| `BASE_COST`          | 12500    |

> | `DATA_WORD_COST`     | 3        |

Reviewer: CREATE and CREATE2 charge `code_deposit_cost = 200 * deployed_code_size` for the cost of storing the code. 

Step 6 does `DATA_WORD_COST * ceil(data_len / 32)`, which appears to be wildelly different (and underpriced ?)

Reviewer: Also note that in case of reconfiguration, we don't pay for the difference in storage footprint. If the new data is the same length, or smaller, than the previous one, we still pay for its storage. Similarly, if the new data is larger than the old one, we don't get any saving from the price we already paid to store the old data.

I'm fine with that.

Author: Yes, I forgot about deployment, I was only thinking of memory expansion.  I'll update those in a separate commit from the simple edits.

Author: set to 200/byte, which matches CREATE

### Target and data layout

`target` is taken from the stack the way every other address operand is, keeping the low 20 bytes and discarding anything above them. Halting on a non-zero high word was considered, on the grounds that a dirty word is never intended, but it would make `SETDELEGATE` the only instruction that treats a stack address more strictly than `CALL` and its siblings do, and callers already assemble addresses knowing the convention.

The data follows the target immediately, with no padding between them. Reserving space for a wider target was also considered and rejected: it would cost 12 bytes of state in every account carrying data, to hold a value that is zero in all of them, hedging a change that may never come.

Because a designator with no trailing data remains exactly 23 bytes, it stays byte-for-byte identical to one written by an EIP-7702 authorization.

### Refusing to overwrite non-designator code

> Step 8 halts when `location` already holds code that does not begin with `DESIGNATOR`. The check draws the line between the two things a caller might be doing. Replacing a designator is reconfiguration, and is the point of the instruction. Replacing anything else means the caller has landed on an account that is not one of its delegations, and there is no case in which continuing is what the caller wanted.

Reviewer: ```suggestion
Step 9 halts when `location` already holds code that does not begin with `DESIGNATOR`. The check draws the line between the two things a caller might be doing. Replacing a designator is reconfiguration, and is the point of the instruction. Replacing anything else means the caller has landed on an account that is not one of its delegations, and there is no case in which continuing is what the caller wanted.
```

> Nothing in the address derivation makes that situation impossible. `location` derives from the executing account and a caller-chosen salt, so a factory computing a salt from untrusted input can be steered onto an address that already holds ordinary contract code, and an address can receive code by other means before the factory reaches it. Without the check, `SETDELEGATE` would silently convert that account into a delegation, discarding its code while leaving its balance, nonce, and storage in place. Halting keeps the instruction unable to destroy anything a caller did not create as a designator, at the cost of one comparison that a correct factory never triggers.

Reviewer: You'd need to manufacture a salt that produce a 160bit collision. Its technically possible, but I would not be worried about it. 

Note that if its possible to colide with a smart contract that has code, its also possible to colide with an EOA that has no code, and in that case the operation will not halt. This would be bad, because the delegation that is set would act as a backdoor that could drain the EOA that it colides with.

Author: code replacement depends on a correct delegate header (current step 9, needs to be empty or start with the delegate designator)

The point I want to make is that people who used to pre-seed addresses based on their own unpublished addresses (counterfactual addresses) will have the same guarantees if they use the same SETDELEGATE factory contract and seed material, which should be the hash of the new PK or however the new address is chosen.  This is highly dependent on the factroy.

Author: I also place this in the same category as a second ECDSA key to an existing address.  About the same difficulty as we are collapsing 256 bytes to 160 via hash.
